res/controllers/Traktor-Kontrol-S3-hid-scripts.js

```diff
@@ -0,0 +1,2227 @@
+        ledValue += TraktorS3.LEDBrightValue;
+    } else {
+        ledValue = 0x00;
+    }
+    this.wheelOutput(group,
+        [ledValue, ledValue, ledValue, ledValue, ledValue, ledValue, ledValue, ledValue]);
+};
+
+TraktorS3.Deck.prototype.wheelOutput = function(group, valueArray) {
+    if (group !== this.activeChannel) {
+        return;
+    }
+
+    for (var i = 0; i < 8; i++) {
+        this.controller.hid.setOutput(this.group, "!wheel" + i, valueArray[i], false);
+    }
+    if (!TraktorS3.batchingOutputs) {
+        for (var packetName in this.controller.hid.OutputPackets) {
+            this.controller.hid.OutputPackets[packetName].send();
+        }
+    }
+};
+
+/////////////////////////
+//// Channel Objects ////
+////
+//// Channels don't have much state, just the fx button state.
+TraktorS3.Channel = function(controller, parentDeck, group) {
+    this.controller = controller;
+    this.parentDeck = parentDeck;
+    this.group = group;
+    this.fxEnabledState = false;
+
+    this.trackDurationSec = 0;
+    this.positionUpdated = false;
+    this.curPosition = -1;
+    this.endOfTrackTimer = 0;
+    this.endOfTrack = false;
+    this.endOfTrackBlinkState = 0;
+    this.vuMeterUpdated = false;
+    this.vuMeterValue = 0;
+
+    this.vuConnection = {};
+    this.clipConnection = {};
+    this.hotcueCallbacks = [];
+};
+
+// Finds the shortest distance between two angles on the wheel, assuming
+// 0-8.0 angle value.
+TraktorS3.wheelSegmentDistance = function(segNum, angle) {
+    // Account for wraparound
+    if (Math.abs(segNum - angle) > 4) {
+        if (angle > segNum) {
+            segNum += 8;
+        } else {
+            angle += 8;
+        }
+    }
+    return Math.abs(angle - segNum);
+};
+
+TraktorS3.Channel.prototype.trackLoadedHandler = function() {
+    var trackSamples = engine.getValue(this.group, "track_samples");
+    if (trackSamples === 0) {
+        this.trackDurationSec = 0;
+        return;
+    }
+    var trackSampleRate = engine.getValue(this.group, "track_samplerate");
+    // Assume stereo.
+    this.trackDurationSec = trackSamples / 2.0 / trackSampleRate;
+};
+
+TraktorS3.Channel.prototype.endOfTrackHandler = function(value) {
+    this.endOfTrack = value;
+    if (!value) {
+        if (this.endOfTrackTimer) {
+            engine.stopTimer(this.endOfTrackTimer);
+            this.endOfTrackTimer = 0;
+        }
+        return;
+    }
+    this.endOfTrackTimer = engine.beginTimer(400, function() {
+        this.endOfTrackBlinkState = !this.endOfTrackBlinkState;
+    }, false);
+};
+
+TraktorS3.Channel.prototype.playpositionChanged = function(value) {
+    if (this.parentDeck.activeChannel !== this.group) {
+        return;
+    }
+
+    // How many segments away from the actual angle should we light?
+    // (in both directions, so "2" will light up to four segments)
+    if (this.trackDurationSec === 0) {
+        var samples = engine.getValue(this.group, "track_samples");
```

Comment:
```suggestion
        var samples = engine.getValue(this.group, "track_loaded");
```
There's a ControlObject specifically for this purpose now.

Reply:
cool, thanks